Add basic cog support (#1590)
urbica/martin: docs/src/sources-cog-files.md

# Cloud Optimized GeoTIFF File Sources

> Martin can also serve raster source like local [COG(Cloud Optimized GeoTIFF)](https://cogeo.org/) files. For cog on remote like S3 and other improvements, you could track them on [issue 875](https://github.com/maplibre/martin/issues/875), we are working on and welcome any assistance.

Reviewer: ```suggestion
Martin can also serve raster sources like local [COG(Cloud Optimized GeoTIFF)](https://cogeo.org/) files. For cog on remote like S3 and other improvements, you could track them on [issue 875](https://github.com/maplibre/martin/issues/875), we are working on and welcome any assistance.
```

# Configured with a combination of directories and dedicated TIFF files.
martin /with/tiff/files /path/to/target.tif
```

## Run Martin with configuration file

```yml
keep_alive: 75

# The socket address to bind [default: 0.0.0.0:3000]
listen_addresses: '0.0.0.0:3000'

# Number of web server workers
worker_processes: 8

# Amount of memory (in MB) to use for caching tiles [default: 512, 0 to disable]
cache_size_mb: 8

# Database configuration. This can also be a list of PG configs.

cog:
  paths:
    # scan this whole dir, matching all *.tif files
    - /dir-path
    # specific TIFF file will be published as a cog source
    - /path/to/target1.tif
    - /path/to/target2.tif
  sources:
    # named source matching source name to a single file
     cog-src1: /path/to/cog1.tif
     cog-src2: /path/to/cog2.tif
```

## About COG

[COG](https://cogeo.org/) is just Cloud Optimized GeoTIFF file.

TIFF is an image file format. TIFF tags are something like key-value pairs to describe the metadata about TIFF file, like `ImageWidth`, `ImageLength`.. etc.

GeoTIFF is a valid TIFF file with a set of TIFF tags to describe "Cartographic" information associated with.

> And COG is a valid GeoTIFF file with some requirements for efficient reading. That is, all COG files are valid GeoTIFF files, but not all GeoTIFF files are valid COG files. For quick access to tiles in TIFF files, Martin relies on these requirements/recommendations(like the order of sub-files, and it must be tiled) so we use the term `COG` over `GeoTIFF` in our documentation and configuration files.

Reviewer: What do you mean by `like the order of sub-files, and it must be tiled`?
Is the following (and => or, and shorter sentences) correct?

```suggestion
And COG is a valid GeoTIFF file with some requirements for efficient reading.
That is, all COG files are valid GeoTIFF files, but not all GeoTIFF files are valid COG files.
For quick access to tiles in TIFF files, Martin relies on these requirements/recommendations (like the order of sub-files, or if it must be tiled).
We use the term `COG` over `GeoTIFF` in our documentation and configuration files.
```

Author: I add some links to doc. You may be interested in [the chapter 7 of COG spec](https://docs.ogc.org/is/21-026/21-026.html#GeoTIFFFormatRequirements) @CommanderStorm

> And COG is a valid GeoTIFF file with some requirements for efficient reading. That is, all COG files are valid GeoTIFF files, but not all GeoTIFF files are valid COG files. For quick access to tiles in TIFF files, Martin relies on these requirements/recommendations(like the order of sub-files, and it must be tiled) so we use the term `COG` over `GeoTIFF` in our documentation and configuration files.

Author: The description about these three and why we use `COG` over `GeoTIFF`  added here. @nyurik

# Cloud Optimized GeoTIFF File Sources

Martin can serve local [COG(Cloud Optimized GeoTIFF)](https://cogeo.org/) files. For cog on remote like S3, you could track it on [issue 875](https://github.com/maplibre/martin/issues/875), we are working on and welcome any assistance.

> A Cloud Optimized GeoTIFF (COG) is a regular GeoTIFF file, aimed at being hosted on a HTTP file server, with an internal organization that enables more efficient workflows on the cloud. It does this by leveraging the ability of clients issuing ​HTTP GET range requests to ask for just the parts of a file they need.

|colory type|bits per sample|supported|status|
|----|----|----|----|
|rgb/rgba|8|✅||
|rgb/rgba|16/32...|🛠️|working on|
|gray|8/16/32...|🛠️|working on|

## Run Martin with CLI to serve cog fiels

```bash
martin /path/to/dir_contains_cog /path/to/cog.tif

> ```

Reviewer: I cannot suggest this, but please split this up to prevent confusing users.
Currently they might infer that `..dir_contains_cog` is nessesary for a `..cog.tif`

`martin /path/to/cog1.tif /path/to/cog2.tif`
`martin /path/to/dir_contains_cog`

Reviewer 2: I would suggest `martin /path/to/dir/with/cog/files` for clarity...

P.S. or `/with/tiff/files` (because `cog` might also be less recognizable)

Author: Made it more detailed now.

# Cloud Optimized GeoTIFF File Sources

> Martin can serve local [COG(Cloud Optimized GeoTIFF)](https://cogeo.org/) files. For cog on remote like S3 and other improvements, you could track them on [issue 875](https://github.com/maplibre/martin/issues/875), we are working on and welcome any assistance.

Reviewer: Could you add a short warning/info section which makes clear that COGs are images and not a vector source? ^^
Also, what do we need to change here https://maplibre.org/martin/using.html

Author: I guess no need to change the Usage and Endpoint APi?

Reviewer 2: Tif or tiff? Seems to have both, but should probably use one?

Author: @nyurik Which one do you perfer? 
[Tiff 6.0 spec](https%3A%2F%2Fwww.itu.int%2Fitudoc%2Fitu-t%2Fcom16%2Ftiff-fx%2Fdocs%2Ftiff6.pdf), [geotiff doc](https://docs.ogc.org/is/19-008r4/19-008r4.html) and [cog doc](https://docs.ogc.org/is/21-026/21-026.html) all use `tiff` as part of their doc name. So we use tiff?

Author: The confusing part is generally the file extension is `tif` and the doc names are `tiff`.

Author: And another to discuss, should we support `*.tiff` and `*.tif` both? Seems they are both right. But I rarely see a `.tiff` file. And if we do support both, do them next PR.

Reviewer 2: I think we should auto-discover both

| colory type | bits per sample | supported | status     |
| ----------- | --------------- | --------- | ---------- |
| rgb/rgba    | 8               | ✅         |            |
| rgb/rgba    | 16/32...        | 🛠️         | working on |
| gray        | 8/16/32...      | 🛠️         | working on |

## Supported compression

* None
* LZW
* Deflate
* PackBits

## Run Martin with CLI to serve cog files

```bash
# Configured with a directory containing TIFF files.
martin /with/tiff/dir1 /with/tiff/dir2
# Configured with dedicated TIFF file.
martin /path/to/target1.tif /path/to/target1.tif
# Configured with a combination of directories and dedicated TIFF files.
martin /with/tiff/files /path/to/target.tif
```

## Run Martin with configuration file

```yml
keep_alive: 75

# The socket address to bind [default: 0.0.0.0:3000]
listen_addresses: '0.0.0.0:3000'

# Number of web server workers
worker_processes: 8

# Amount of memory (in MB) to use for caching tiles [default: 512, 0 to disable]
cache_size_mb: 8

# Database configuration. This can also be a list of PG configs.

cog:
  paths:
    # scan this whole dir, matching all *.tif files
    - /dir-path

>     # specific tif file will be published as a cog source

Reviewer 2: ```suggestion
    # specific TIFF file will be published as a cog source
```

    - /path/to/target1.tif
    - /path/to/target2.tif
  sources:
    # named source matching source name to a single file
     cog-src1: /path/to/cog1.tif
     cog-src2: /path/to/cog2.tif
```

## About COG

[COG](https://cogeo.org/) is just Cloud Optimized GeoTIFF file. You could generate cog with `gdal_translate` or `gdalwarp`. See more details in [gdal doc](https://gdal.org/en/latest/drivers/raster/cog.html).

```bash
# gdal-bin installation
# sudo apt update
# sudo apt install gdal-bin

# gdalwarp
gdalwarp src1.tif src2.tif out.tif -of COG

# or gdal_translate
gdal_translate input.tif output_cog.tif -of COG
```

### The mapping from ZXY to tiff chunk

> * A single tif file could contains many subfile about same spatial area, each has different resollution

Reviewer 2: ```suggestion
* A single TIFF file could contains many sub-file about same spatial area, each has different resolution
```

* A sub file is organized with many tiles

> So basically there's a mapping from zxy to tile of subfile of tif.

Reviewer 2: ```suggestion
So basically there's a mapping from zxy to tile of sub-file of TIFF.
```

| zxy        | mapping to                |
| ---------- | ------------------------- |

> | Zoom level | which subfile in tif file |

Reviewer 2: ```suggestion
| Zoom level | which sub-file in TIFF file |
```
